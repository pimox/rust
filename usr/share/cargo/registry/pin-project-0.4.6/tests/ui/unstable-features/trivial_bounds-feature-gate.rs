// NB: If you change this test, change 'trivial_bounds.rs' at the same time.

use std::marker::{PhantomData, PhantomPinned};

fn phantom_pinned() {
    struct Foo(PhantomPinned);

    impl Unpin for Foo where PhantomPinned: Unpin {} //~ ERROR E0277

    struct Wrapper<T>(T);

    impl<T> Unpin for Wrapper<T> where T: Unpin {}

    struct Bar(PhantomPinned);

    impl Unpin for Bar where Wrapper<PhantomPinned>: Unpin {} //~ ERROR E0277

    struct WrapperWithLifetime<'a, T>(PhantomData<&'a ()>, T);

    impl<T> Unpin for WrapperWithLifetime<'_, T> where T: Unpin {}

    struct Baz(PhantomPinned);

    impl<'a> Unpin for Baz where WrapperWithLifetime<'a, PhantomPinned>: Unpin {}
    // Ok
}

fn inner() {
    struct Inner(PhantomPinned);

    struct Foo(Inner);

    impl Unpin for Foo where Inner: Unpin {} //~ ERROR E0277

    struct Wrapper<T>(T);

    impl<T> Unpin for Wrapper<T> where T: Unpin {}

    struct Bar(Inner);

    impl Unpin for Bar where Wrapper<Inner>: Unpin {} //~ ERROR E0277

    struct WrapperWithLifetime<'a, T>(PhantomData<&'a ()>, T);

    impl<T> Unpin for WrapperWithLifetime<'_, T> where T: Unpin {}

    struct Baz(Inner);

    impl<'a> Unpin for Baz where WrapperWithLifetime<'a, Inner>: Unpin {} // Ok
}

fn main() {}
