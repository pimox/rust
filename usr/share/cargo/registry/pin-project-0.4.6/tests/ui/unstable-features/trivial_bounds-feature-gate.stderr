error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied
 --> $DIR/trivial_bounds-feature-gate.rs:8:5
  |
8 |     impl Unpin for Foo where PhantomPinned: Unpin {} //~ ERROR E0277
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
  |
  = help: the following implementations were found:
            <std::marker::PhantomPinned as std::marker::Unpin>
  = help: see issue #48214
  = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied
  --> $DIR/trivial_bounds-feature-gate.rs:16:5
   |
16 |     impl Unpin for Bar where Wrapper<PhantomPinned>: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required because of the requirements on the impl of `std::marker::Unpin` for `phantom_pinned::Wrapper<std::marker::PhantomPinned>`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied in `inner::Inner`
  --> $DIR/trivial_bounds-feature-gate.rs:33:5
   |
33 |     impl Unpin for Foo where Inner: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `inner::Inner`, the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required because it appears within the type `inner::Inner`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied in `inner::Inner`
  --> $DIR/trivial_bounds-feature-gate.rs:41:5
   |
41 |     impl Unpin for Bar where Wrapper<Inner>: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `inner::Inner`, the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required because it appears within the type `inner::Inner`
   = note: required because of the requirements on the impl of `std::marker::Unpin` for `inner::Wrapper<inner::Inner>`
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
